### PR TITLE
do viaq processing/formatting in ruby code; create index names

### PR DESCRIPTION
**Pull request by richm**
This allows the viaq filter to be used to convert systemd journal
records read in by the systemd journal input.  This also allows
the viaq filter to construct Elasticsearch index names for
openshift/origin-aggregated-logging for the operations and
kubernetes indices.

### CODE REVIEW

**Comment by richm** [comment]
Here is what the openshift logging config looks like with this change:
- filter-syslog-record-transformer.conf and filter-k8s-record-transformer.conf are empty: https://github.com/richm/origin-aggregated-logging/blob/test-viaq-e3d6f25-support-both-index-names-and-prefixes/fluentd/configs.d/openshift/filter-k8s-record-transform.conf
- filter-viaq-data-model.conf is expanded: https://github.com/richm/origin-aggregated-logging/blob/test-viaq-e3d6f25-support-both-index-names-and-prefixes/fluentd/configs.d/openshift/filter-viaq-data-model.conf
- the elasticsearch output plugins are now type `elasticsearch` and use the `target_index_key` for the index name: https://github.com/richm/origin-aggregated-logging/blob/test-viaq-e3d6f25-support-both-index-names-and-prefixes/fluentd/configs.d/openshift/output-es-ops-config.conf#L6
@portante @jcantrill PTAL

File: lib/fluent/plugin/filter_viaq_data_model.rb [hunk 1]
```diff
@@ -15,8 +15,39 @@
...
+        Fluent::MatchPattern.create(str)
+      }
+      if patterns.length == 1
+        @pattern = patterns[0]
+      else
+        @pattern = Fluent::OrMatchPattern.new(patterns)
+      end
+    end
+    def match(tag)
+      if @pattern.match(tag)
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_viaq_data_model.rb]: Looking at 'match' i think you could just return the result of `@pattern.match(tag)`

File: lib/fluent/plugin/filter_viaq_data_model.rb [hunk 2]
```diff
@@ -76,6 +151,43 @@ def configure(conf)
...
+        @formatters.each do |fmtr|
+          matcher = ViaqMatchClass.new(fmtr.tag, nil)
+          fmtr.instance_eval{ @params[:matcher] = matcher }
+          fmtr.instance_eval{ @params[:fmtr_type] = fmtr.type }
+          if fmtr.remove_keys
+            fmtr.instance_eval{ @params[:fmtr_remove_keys] = fmtr.remove_keys.split(',') }
+          else
+            fmtr.instance_eval{ @params[:fmtr_remove_keys] = nil }
+          end
+          if fmtr.type == :sys_journal || fmtr.type == :k8s_journal
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_viaq_data_model.rb]: Good opportunity to use a 'case' but not required.

File: lib/fluent/plugin/filter_viaq_data_model.rb [hunk 3]
```diff
@@ -104,12 +216,244 @@ def delempty(thing)
       thing
     end
 
+    # map of journal fields to viaq data model field
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_viaq_data_model.rb]: Consider moving these into a separate module and mixing in to reduce file size

File: lib/fluent/plugin/filter_viaq_data_model.rb [hunk 4]
```diff
@@ -104,12 +216,244 @@ def delempty(thing)
...
+      "_KERNEL_DEVICE"    => "KERNEL_DEVICE",
+      "_KERNEL_SUBSYSTEM" => "KERNEL_SUBSYSTEM",
+      "_UDEV_SYSNAME"     => "UDEV_SYSNAME",
+      "_UDEV_DEVNODE"     => "UDEV_DEVNODE",
+      "_UDEV_DEVLINK"     => "UDEV_DEVLINK",
+    }
+
+    JOURNAL_TIME_FIELDS = ['_SOURCE_REALTIME_TIMESTAMP', '__REALTIME_TIMESTAMP']
+
+    def process_journal_fields(tag, time, record, fmtr_type)
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_viaq_data_model.rb]: I don't know how easy this class is to test, but consider moving it to a module to be mixed in.  In may facilitate unit testing

File: lib/fluent/plugin/filter_viaq_data_model.rb [hunk 5]
```diff
@@ -104,12 +216,244 @@ def delempty(thing)
...
+      end
+      begin
+        pri_index = ('%d' % record['PRIORITY'] || 9).to_i
+        if pri_index < 0
+          pri_index = 9
+        elsif pri_index > 9
+          pri_index = 9
+        end
+      rescue
+        pri_index = 9
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_viaq_data_model.rb]: We get here if PRIORITY is non_numeric?

File: lib/fluent/plugin/filter_viaq_data_model.rb [hunk 6]
```diff
@@ -104,12 +216,244 @@ def delempty(thing)
...
+      record['level'] = ["emerg", "alert", "crit", "err", "warning", "notice", "info", "debug", "trace", "unknown"][pri_index]
+      JOURNAL_TIME_FIELDS.each do |field|
+        if record[field]
+          record['time'] = Time.at(record[field].to_f / 1000000.0).utc.to_datetime.rfc3339(6)
+          break
+        end
+      end
+      if fmtr_type == :sys_journal
+        record['message'] = record['MESSAGE']
+        if record['_HOSTNAME'].eql?('localhost') && @docker_hostname
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_viaq_data_model.rb]: Do we need to invert this incase the recored does not have hostname.. `'localhost'.eql?(record['_HOSTNAME'])`

File: lib/fluent/plugin/filter_viaq_data_model.rb (hunk 5 above)
Comment (richm) [reply lib/fluent/plugin/filter_viaq_data_model.rb]: yes

File: lib/fluent/plugin/filter_viaq_data_model.rb (hunk 6 above)
Comment (richm) [reply lib/fluent/plugin/filter_viaq_data_model.rb]: no - if no `_HOSTNAME` then `record['_HOSTNAME']` is `nil` and `nil.eql?('localhost')` is `false`

**Comment by richm** [comment]
@jcantrill thanks - ptal at latest commit

**Review by jcantrill — APPROVED** [review]

**Pull request merged**